### ISSUE TITLE
ctc_beam_search_decoder()'s log_probabilities holds invalid values

### ISSUE THREAD

**Issue by kdavis-mozilla** [tensorflow/tensorflow#6034]
### Environment info
Operating System: OS X 10.11.6
TF Version: 0.10.0rc0 (No GPU)

### Example
Ran `log_probabilities` op created from
```python
decoded, log_probabilities = ctc_ops.ctc_beam_search_decoder(logits, \
                                                             seq_length, \
                                                             beam_width=beam_width, \
                                                             top_paths=top_paths, \
                                                             merge_repeated=False)
```
The `decoded` result is as expected. However, the `log_probabilities` contains positive values which can not be log probabilities. For example, with batch size 4 and top_paths=10 the `log_probabilities` printout is as follows
```
[[ 3.85424066 -1.97321272 -1.99056399 -2.18253303 -2.18592954 -2.40727925
  -2.87798476 -2.88267159 -2.94563317 -2.94854331]
 [ 3.85424066 -1.97321272 -1.99056399 -2.18253303 -2.18592954 -2.40727925
  -2.87798476 -2.88267159 -2.94563317 -2.94854331]
 [ 3.85424066 -1.97321272 -1.99056399 -2.18253303 -2.18592954 -2.40727925
  -2.87798476 -2.88267159 -2.94563317 -2.94854331]
 [ 3.85424066 -1.97321272 -1.99056399 -2.18253303 -2.18592954 -2.40727925
  -2.87798476 -2.88267159 -2.94563317 -2.94854331]]
```

### Other attempted solutions
None

### Logs or other output that would be helpful
Link to the entire code in context[[1](https://github.com/mozilla/DeepSpeech/blob/issue8++/DeepSpeech.ipynb)]

**Comment by ebrevdo** [comment]
I may not get to this before January.

**Comment by kdavis-mozilla** [comment]
Is the returned `log_probabilities` supposed to represented the logs of a probabilities, as documented[[1](https://www.tensorflow.org/versions/r0.12/api_docs/python/nn.html#ctc_beam_search_decoder)], or does it represent _un-normalized_  logs of a probabilities?

**Comment by ebrevdo** [comment]
Should be normalized, I'll have to trace this exact example to see what's going on

**Comment by vrenkens** [comment]
I have the same problem, with beam search I get positive scores but with greedy I get negative ones

**Comment by kdavis-mozilla** [comment]
@ebrevdo Has there been any progress on this?

**Comment by ebrevdo** [comment]
Please provide a small example input that causes the issue so I can debug

**Comment by vrenkens** [comment]
I am not really familiar with the code, but is it possible that it is caused by line 183 in tensorflow/tensorflow/core/util/ctc/ctc_beam_search.h: (*scores)(b, i) = -beam_log_probabilities[i]; ?

**Comment by kdavis-mozilla** [comment]
@ebrevdo Until I get time to write a smaller, example one can look at [ctc_decoder_ops_test.py](https://github.com/tensorflow/tensorflow/blob/584d4921014db921a8f4722749adff09737826a8/tensorflow/python/kernel_tests/ctc_decoder_ops_test.py)

Documentation of [`ctc_beam_search_decoder`](https://www.tensorflow.org/api_docs/python/nn/connectionist_temporal_classification__ctc_#ctc_beam_search_decoder) states
```
...
Returns:

A tuple (decoded, log_probabilities) where

decoded: A list of length top_paths, where decoded[j] is a SparseTensor ...
log_probability: A float matrix (batch_size x top_paths) containing sequence log-probabilities.
```

However, for [ctc_decoder_ops_test.py](https://github.com/tensorflow/tensorflow/blob/584d4921014db921a8f4722749adff09737826a8/tensorflow/python/kernel_tests/ctc_decoder_ops_test.py) the expected results, the `log_prob_truth` values,  are
```
    log_prob_truth = np.array(
        [
            0.584855,  # output beam 0
            0.389139  # output beam 1
        ],
        np.float32)[np.newaxis, :]
```
These are not log probabilities as they are positive.

**Comment by Garvys** [comment]
Is there any progress on this issue ? I have this problem on TF 1.0

**Comment by tychotax** [comment]
@ebrevdo same issue here!

**Comment by pdcoded** [comment]
@ebrevdo any progress on this issue? any workaround?

**Comment by qiqiguaitm** [comment]
same issue

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by BradNeuberg** [comment]
This issue actually prevented a project here at Dropbox from migrating our training pipeline from Torch to TensorFlow, unfortunately causing us to stay with Torch.

**Comment by ebrevdo** [comment]
@kdavis-mozilla in your initial report, what were the seq_len values passed in?

**Comment by kdavis-mozilla** [comment]
@ebrevdo Sorry but I don't even know anymore; it was over a year ago.

We've since created our own custom `ctc_beam_search_decoder` operator extending the TF operator with KenLM. We use that now.

As far a I remember, the problem is still there. The relative ordering of results via their `log_probabilities` is fine, but the absolute `log_probability` values are not.

**Comment by ryanleary** [comment]
The problem is still there. It's related to the subtraction of the `maxCoeff` from the input and then the subsequent negation of the log probs. I can submit a PR with the fix, but don't have time to workthrough the broken test cases.

**Comment by ebrevdo** [comment]
The subtraction is there to avoid overflow/underflow issues iirc.  Is your
suggestion to add the values back in at the very end?

**Comment by chenjiasheng** [comment]
@ryanleary
Could you please guide me, how to fix this issue in Python side, with touching C++?

top=paths=3, beam_width=400, I get these log-probs:
```
array([[ 0.09569344, -2.35573864, -3.55027127],
       [ 0.31278262, -0.11559263, -0.22166124],
       [ 0.01575552, -0.69348544, -0.84170169],
       [ 0.08385472, -0.10281177, -1.23370361],
       [ 1.34617054,  1.31623089, -0.14258756],
       [ 0.1471574 , -0.13724545, -0.34559944],
       [ 0.3714101 ,  0.02229187, -0.05457714],
       [ 0.72338897, -0.42095903, -0.4398351 ],
       [ 0.66841668, -0.19791916, -0.73667264],
       [ 0.11281018, -0.14468569, -2.61785507],
       [ 0.02421612, -0.29929882, -0.64302784],
       [ 0.01053159, -0.2815102 , -1.38790202],
       [ 0.03269538, -0.30419388, -0.4501299 ],
       [ 0.00933365, -0.72536844, -1.57513046],
       [ 0.40954223,  0.0353958 , -1.72318888],
       ............
      ], dtype=float32)
```

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by ryanleary** [comment]
@ebrevdo yes, I think something like that needs to occur (to the extent that it can be done). What underflow/overflow issues have you experienced? Is that still really an issue even though we're in the log prob space?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 92 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 107 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by shokou-chin** [comment]
Is there any progress on this issue?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by igormq** [comment]
Up.

**Comment by ebrevdo** [comment]
Sorry; unfortunately i do not currently have cycles to look at this.  It's
an important issue though, so I won't close it for now.

**Comment by igormq** [comment]
What do you mean?

This issue has more than one year now, and as you said is an important issue. Isn't there anyone here that can have a look into this?

**Comment by ebrevdo** [comment]
It's being triaged

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 34 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by ebrevdo** [comment]
I think this is fixed now?

**Comment by kdavis-mozilla** [comment]
@ebrevdo What commit fixed this problem?

**Comment by ebrevdo** [comment]
https://github.com/tensorflow/tensorflow/pull/21187
